File: 2023/day07/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <cstdio>
#include <cstring>
#include <vector>
#include <algorithm>

constexpr int HAND_SIZE = 5;
constexpr int NUM_CARDS = 13;

// Hand type rankings (higher value = stronger hand)
enum HandType {
    HIGH_CARD = 0,
    ONE_PAIR = 1,
    TWO_PAIR = 2,
    THREE_OF_A_KIND = 3,
    FULL_HOUSE = 4,
    FOUR_OF_A_KIND = 5,
    FIVE_OF_A_KIND = 6
};

struct Hand {
    char cards[HAND_SIZE + 1];
    int bid;
    int type;
    int card_values[HAND_SIZE];
};

// Card strength order for Part 1 (higher index = stronger)
static const char* CARD_STRENGTH = "23456789TJQKA";
// Card strength order for Part 2 (J is weakest)
static const char* CARD_STRENGTH_JOKER = "J23456789TQKA";

inline int cardValue(char c, const char* strength_order) {
    const char* p = std::strchr(strength_order, c);
    return p ? static_cast<int>(p - strength_order) : -1;
}

int getHandType(const char* hand) {
    int counts[NUM_CARDS] = {0};

    // Count occurrences of each card
    for (int i = 0; i < HAND_SIZE; i++) {
        int idx = cardValue(hand[i], CARD_STRENGTH);
        if (idx >= 0) counts[idx]++;
    }

    // Get sorted counts (descending)
    int sorted_counts[5] = {0};
    int num_counts = 0;
    for (int i = 0; i < NUM_CARDS; i++) {
        if (counts[i] > 0) {
            sorted_counts[num_counts++] = counts[i];
        }
    }

    // Simple bubble sort (small array)
    for (int i = 0; i < num_counts - 1; i++) {
        for (int j = i + 1; j < num_counts; j++) {
            if (sorted_counts[j] > sorted_counts[i]) {
                std::swap(sorted_counts[i], sorted_counts[j]);
            }
        }
    }

    // Determine hand type based on sorted counts
    if (num_counts == 1) return FIVE_OF_A_KIND;
    if (num_counts == 2 && sorted_counts[0] == 4) return FOUR_OF_A_KIND;
    if (num_counts == 2 && sorted_counts[0] == 3) return FULL_HOUSE;
    if (num_counts == 3 && sorted_counts[0] == 3) return THREE_OF_A_KIND;
    if (num_counts == 3 && sorted_counts[0] == 2) return TWO_PAIR;
    if (num_counts == 4) return ONE_PAIR;
    return HIGH_CARD;
}

int getHandTypeWithJokers(const char* hand) {
    int joker_count = 0;
    for (int i = 0; i < HAND_SIZE; i++) {
        if (hand[i] == 'J') joker_count++;
    }

    if (joker_count == 0) return getHandType(hand);
    if (joker_count == 5) return FIVE_OF_A_KIND;

    // Count non-joker cards
    int counts[NUM_CARDS] = {0};
    for (int i = 0; i < HAND_SIZE; i++) {
        if (hand[i] != 'J') {
            int idx = cardValue(hand[i], CARD_STRENGTH);
            if (idx >= 0) counts[idx]++;
        }
    }

    // Get sorted counts (descending)
    int sorted_counts[5] = {0};
    int num_counts = 0;
    for (int i = 0; i < NUM_CARDS; i++) {
        if (counts[i] > 0) {
            sorted_counts[num_counts++] = counts[i];
        }
    }

    // Simple bubble sort
    for (int i = 0; i < num_counts - 1; i++) {
        for (int j = i + 1; j < num_counts; j++) {
            if (sorted_counts[j] > sorted_counts[i]) {
                std::swap(sorted_counts[i], sorted_counts[j]);
            }
        }
    }

    // Add jokers to highest count
    sorted_counts[0] += joker_count;

    // Determine hand type based on sorted counts
    if (num_counts == 1 || sorted_counts[0] == 5) return FIVE_OF_A_KIND;
    if (sorted_counts[0] == 4) return FOUR_OF_A_KIND;
    if (sorted_counts[0] == 3 && sorted_counts[1] == 2) return FULL_HOUSE;
    if (sorted_counts[0] == 3) return THREE_OF_A_KIND;
    if (sorted_counts[0] == 2 && sorted_counts[1] == 2) return TWO_PAIR;
    if (sorted_counts[0] == 2) return ONE_PAIR;
    return HIGH_CARD;
}

bool compareHands(const Hand& a, const Hand& b) {
    // Compare by type first
    if (a.type != b.type) {
        return a.type < b.type;
    }

    // Then by card values left to right
    for (int i = 0; i < HAND_SIZE; i++) {
        if (a.card_values[i] != b.card_values[i]) {
            return a.card_values[i] < b.card_values[i];
        }
    }
    return false;
}

long long solve(std::vector<Hand>& hands, bool part2) {
    // Precompute type and card values for each hand
    const char* strength_order = part2 ? CARD_STRENGTH_JOKER : CARD_STRENGTH;

    for (auto& hand : hands) {
        hand.type = part2 ? getHandTypeWithJokers(hand.cards) : getHandType(hand.cards);
        for (int j = 0; j < HAND_SIZE; j++) {
            hand.card_values[j] = cardValue(hand.cards[j], strength_order);
        }
    }

    // Sort hands
    std::sort(hands.begin(), hands.end(), compareHands);

    // Calculate total winnings
    long long total = 0;
    int rank = 1;
    for (const auto& hand : hands) {
        total += static_cast<long long>(rank++) * hand.bid;
    }

    return total;
}

int main() {
    FILE* fp = std::fopen("../input.txt", "r");
    if (!fp) {
        std::cerr << "Could not open input.txt\n";
        return 1;
    }

    std::vector<Hand> hands1;
    std::vector<Hand> hands2;
    hands1.reserve(1024);
    hands2.reserve(1024);

    char line[256];
    while (std::fgets(line, sizeof(line), fp)) {
        Hand hand;
        if (std::sscanf(line, "%5s %d", hand.cards, &hand.bid) == 2) {
            hands1.push_back(hand);
            hands2.push_back(hand);
        }
    }
    std::fclose(fp);

    long long part1 = solve(hands1, false);
    long long part2_result = solve(hands2, true);

    std::cout << "Part 1: " << part1 << '\n';
    std::cout << "Part 2: " << part2_result << '\n';

    return 0;
}

File: 2023/day07/zig/solution.zig
const std = @import("std");

const Hand = struct {
    cards: [5]u8,
    bid: u32,
    hand_type: u8,
    card_values: [5]u8,
};

fn cardStrength(card: u8, with_jokers: bool) u8 {
    if (with_jokers) {
        return switch (card) {
            'J' => 0,
            '2' => 1,
            '3' => 2,
            '4' => 3,
            '5' => 4,
            '6' => 5,
            '7' => 6,
            '8' => 7,
            '9' => 8,
            'T' => 9,
            'Q' => 10,
            'K' => 11,
            'A' => 12,
            else => 0,
        };
    } else {
        return switch (card) {
            '2' => 0,
            '3' => 1,
            '4' => 2,
            '5' => 3,
            '6' => 4,
            '7' => 5,
            '8' => 6,
            '9' => 7,
            'T' => 8,
            'J' => 9,
            'Q' => 10,
            'K' => 11,
            'A' => 12,
            else => 0,
        };
    }
}

fn getHandType(cards: [5]u8, with_jokers: bool) u8 {
    var counts: [13]u8 = .{0} ** 13;
    var joker_count: u8 = 0;

    for (cards) |card| {
        if (with_jokers and card == 'J') {
            joker_count += 1;
        } else {
            counts[cardStrength(card, false)] += 1;
        }
    }

    // Sort counts descending
    std.mem.sort(u8, &counts, {}, std.sort.desc(u8));

    // Add jokers to the highest count
    if (with_jokers) {
        counts[0] += joker_count;
    }

    // Determine hand type
    if (counts[0] == 5) {
        return 6; // Five of a kind
    } else if (counts[0] == 4) {
        return 5; // Four of a kind
    } else if (counts[0] == 3 and counts[1] == 2) {
        return 4; // Full house
    } else if (counts[0] == 3) {
        return 3; // Three of a kind
    } else if (counts[0] == 2 and counts[1] == 2) {
        return 2; // Two pair
    } else if (counts[0] == 2) {
        return 1; // One pair
    } else {
        return 0; // High card
    }
}

fn compareHands(context: void, a: Hand, b: Hand) bool {
    _ = context;
    if (a.hand_type != b.hand_type) {
        return a.hand_type < b.hand_type;
    }
    for (0..5) |i| {
        if (a.card_values[i] != b.card_values[i]) {
            return a.card_values[i] < b.card_values[i];
        }
    }
    return false;
}

fn parseHands(allocator: std.mem.Allocator, input: []const u8, with_jokers: bool) ![]Hand {
    var hands: std.ArrayListUnmanaged(Hand) = .empty;
    errdefer hands.deinit(allocator);

    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;

        var parts = std.mem.splitScalar(u8, line, ' ');
        const hand_str = parts.next() orelse continue;
        const bid_str = parts.next() orelse continue;

        if (hand_str.len != 5) continue;

        var hand: Hand = undefined;
        hand.bid = std.fmt.parseInt(u32, bid_str, 10) catch continue;

        for (0..5) |i| {
            hand.cards[i] = hand_str[i];
            hand.card_values[i] = cardStrength(hand_str[i], with_jokers);
        }

        hand.hand_type = getHandType(hand.cards, with_jokers);
        try hands.append(allocator, hand);
    }

    return hands.toOwnedSlice(allocator);
}

fn calculateWinnings(hands: []Hand) u64 {
    std.mem.sort(Hand, hands, {}, compareHands);

    var total: u64 = 0;
    for (hands, 1..) |hand, rank| {
        total += rank * hand.bid;
    }
    return total;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const input = try std.fs.cwd().readFileAlloc(allocator, "../input.txt", 1024 * 1024);
    defer allocator.free(input);

    // Part 1
    const hands1 = try parseHands(allocator, input, false);
    defer allocator.free(hands1);
    const part1 = calculateWinnings(hands1);

    // Part 2
    const hands2 = try parseHands(allocator, input, true);
    defer allocator.free(hands2);
    const part2 = calculateWinnings(hands2);

    std.debug.print("Part 1: {d}\n", .{part1});
    std.debug.print("Part 2: {d}\n", .{part2});
}
